src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java

```diff
@@ -0,0 +1,97 @@
+     * @param queryTopDocs query results that need to be normalized, mutated by method execution
+     * @param combinationTechnique exact combination method that should be applied
+     * @return list of max combined scores for each shard
+     */
+    public List<Float> combineScores(final CompoundTopDocs[] queryTopDocs, final ScoreCombinationTechnique combinationTechnique) {
+        List<Float> maxScores = new ArrayList<>();
+        for (int i = 0; i < queryTopDocs.length; i++) {
+            CompoundTopDocs compoundQueryTopDocs = queryTopDocs[i];
+            if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.totalHits.value == 0) {
+                maxScores.add(ZERO_SCORE);
+                continue;
+            }
+            List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+            int shardId = compoundQueryTopDocs.scoreDocs[0].shardIndex;
+            Map<Integer, float[]> normalizedScoresPerDoc = new HashMap<>();
+            int maxHits = 0;
+            TotalHits.Relation totalHits = TotalHits.Relation.EQUAL_TO;
+            for (int j = 0; j < topDocsPerSubQuery.size(); j++) {
+                TopDocs topDocs = topDocsPerSubQuery.get(j);
+                int hits = 0;
+                for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                    if (!normalizedScoresPerDoc.containsKey(scoreDoc.doc)) {
+                        float[] scores = new float[topDocsPerSubQuery.size()];
+                        // we initialize with -1.0, as after normalization it's possible that score is 0.0
+                        Arrays.fill(scores, -1.0f);
+                        normalizedScoresPerDoc.put(scoreDoc.doc, scores);
+                    }
+                    normalizedScoresPerDoc.get(scoreDoc.doc)[j] = scoreDoc.score;
+                    hits++;
+                }
+                maxHits = Math.max(maxHits, hits);
+            }
+            if (topDocsPerSubQuery.stream()
+                .anyMatch(topDocs -> topDocs.totalHits.relation == TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO)) {
+                totalHits = TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO;
+            }
+            Map<Integer, Float> combinedNormalizedScoresByDocId = normalizedScoresPerDoc.entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, entry -> combinationTechnique.combine(entry.getValue())));
+            // create priority queue, make it max heap by the score
+            PriorityQueue<Integer> pq = new PriorityQueue<>(
+                (a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a))
+            );
+            // we're merging docs with normalized and combined scores. we need to have only maxHits results
+            for (int docId : normalizedScoresPerDoc.keySet()) {
+                pq.add(docId);
+            }
```

Comment:
can we use HitQueue interface provided by lucene here? rather than using our own PQ ?

Reply:
I'm not sure we'll benefit from that class. It's simplifies  work with heap in case elements are updated after being added to the heap, but it's not our case. Other than that it's same standard java PriorityQueue. Also it's designed for ScoreDoc instances, but we're operating with scores.